Devel/halfway gate

MichKe95 opened the pull request:
Fixes #29 and Fixes #30 .

**Summary:**
- Add new Device class
- Add flux pulse amplitude, T1 and T2 swap spectroscopy
- Preparations for using HDAWG with internal modulation
- General bug fixes and improvements

**Changes proposed in this pull request:**
- introduce device object:
   - stores two qubit gate parameters
   - runs multi-qubit experiments and calibration routines 
- timedomain_analysis.py:
   - add analyses for flux-pulse amplitude, and T1 and T2 swap spectroscopy
   - add new flag in options dictionary that allows for running a PCA over the whole data set 
     (global_PCA=True)
- fitting_models.py:
   - **adapt Qubit_dac_to_freq and Qubit_freq_to_dac for swap spectroscopy (set E_c=0)**
- QuDev_transmon.py
   - add flux pulse parameters that store the parameters for the flux pulse of the qubit (used e.g.
      in flux pulse scope)
   - add measure_flux_pulse_amplitude, measure_T1_freq_sweep (swap spectroscopy) and 
      measure_T2_freq_sweep
- fluxing_sequences.py:
   - adapt dynamic_phase_seq to allow for FLIP gate
   - adapt chevron_seqs to allow for FLIP gate
   - add num_cz_gates parameter to chevron_seqs
   - fix bug in flux_pulse_scope_seq that resulted in the flux pulse not starting at delay=0 but at
     delay=-buffer_length_start
   - add flux_pulse_amplitude_seq
   - add T2_freq_sweep_seq
   - add T1_freq_sweep_seq
   - adapt cphase_seq to allow for use of FLIP gate
   - **fix #29 and #30** 
- single_qubit_tek_seq_elts.py
   - **adapt ramsey_active_reset sequence to put the pi half pulses into different elements if
      possible**  
- pulsar.py
   - add possibility of using the internal modulation capabilities for the hdawg
- pulse.py
   - add method pulse_params
- pulse_library.py
   - add BufferedFLIPPulse and BufferedNZFLIPPulse
   - adapt BufferedCZPulse to return waveforms starting with 0
- segment.py
   - adapt calculate_hash to allow for internal modulation capabilities

Inviting @antsr to review.

FYI @stephlazar @christiankraglundandersen  @chellings  @nathlacroix  @jo3141

MichKe95 committed:
Add BufferedHalfwayPulse as new type of two qubit pulse.
Add fluxpulse_amplitude_sequence used to determine the qubit frequency
in dependence of fluxpulse amplitude.
Add measure_flux_pulse_amplitude to run a fluxpulse amplitude measurement.

MichKe95 committed:
Bug fix: add ch_qb1 and ch_qb2 to QuDev_transmon method add_CZ_pulse.

Commit:
Removes check for ziRuntimeError in ZI_HDAWG_core.

MichKe95 committed:
Fix bug with channel parameter for CZ pulse. Now use channel and channel2 for the (possible two channels where the flux pulse should be played).

MichKe95 committed:
Merge branch 'devel/halfway_gate' of https://github.com/QudevETH/PycQED_py3 into devel/halfway_gate

MichKe95 committed:
Add decay_freq_seq sequence to determine T1 for different qubit frequencies.

MichKe95 committed:
Add FluxAmplitudeSweepAnalysis that finds the fit parameters for the frequency vs amplitude plot.

QuDev_transmon can update the ge_freq_fit_para dictionary with the fit parameters.

User can specify both amplitudes or frequencies for the measure_T1_freq_sweep for the sweep.

MichKe95 committed:
FluxAmplitudeSweepAnalysis improvements:
* fitting works more reliably
* add functionality to manually mask data taht should not be considered
* fitting parameters now added as qubit parameter

MichKe95 committed:
Add T1FrequencySweepAnalysis:
* outputs a plot of T1 in dependence of frequency/amplitude of the flux pulse.

MichKe95 committed:
segment.py
* Fix bug that caused waveforms not to be multiples of granularity

fluxing_sequences.py
* make RO sequence more economical using repeat

timedomain_analysis.py
* general bug fixes
* add ability for program to automatically remove fit points with large stderr.

Commit:
Fix bug that caused to data to be rotated twice when using calibration points.

antsr committed:
Add T2FrequencySweep.
Fix bug with detector_functions.py that did not correctly normalize UHFQA results.

MichKe95 committed:
Add measure_T2_freq_sweep
* Performs Ramsey experiment with interleaved flux pulse
* Sweeps over flux pulse amplitude, flux pulse length and phase of difference
* Analysis determines T2 for each amplitude

antsr committed:
Debug T2FrequencySweepAnalysis.

MichKe95 committed:
Improve plots of T1FrequencySweepAnalysis.

MichKe95 committed:
Adapt pulsar to allow for waveform modulation on HDAWG devices.

MichKe95 committed:
Modify pulsar to allow for programming HDAWG with modulated waveforms.

MichKe95 committed:
Bug fixes in the methods used for programming the HDAWG using internal modulation.

MichKe95 committed:
Add Device class V1
- stores chip parameters (e.g connectivity, two-qubit gate params)
- runs multi-qubit experiments

Commit:
Merge branch 'Proj/XLD' into devel/halfway_gate

# Conflicts:
#	pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py

Commit:
Added features:
- add flux_buffer_length for BufferedHalfwayPulse and BufferedNZHalfwayPulse that results
  in qubits not being fluxed at the same time but one after the other

Bug fixes:
- Debugcphase and chevron sequence to handle cases where ge pulses have different length
- Fix bugs with FluxAmplitudeSweepAnalysis
- Fix bugs with measure_chevron and measure_cphase in device.py
- general bug fixes and stability improvements

Commit:
Merge branch 'qudev_master' into devel/halfway_gate

# Conflicts:
#	pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py

Commit:
Merge branch 'devel/analysis_v3' into devel/halfway_gate

Commit:
Add features:
* Add RB and dynamic phase measurement and tomography to device class.

General bug fixes:
* fix bug in tomography sequence

Commit:
- Fix single shot readout when updating state_prob_mtx for readout correction

Commit:
Merge branch 'qudev_master' into devel/halfway_gate

# Conflicts:
#	pycqed/analysis/analysis_toolbox.py
#	pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
#	pycqed/measurement/detector_functions.py
#	pycqed/measurement/pulse_sequences/fluxing_sequences.py
#	pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
#	pycqed/measurement/waveform_control/pulse_library.py
#	pycqed/measurement/waveform_control/segment.py

MichKe95 committed:
* import of MAX_PATH raises ValueError for macOS systems

MichKe95 committed:
* pulse_library.py: call Pulse via pulse.py in HalfwayPulse
* segment.py: clean up formatting issue after merge

MichKe95 committed:
* clean up device.py file
* change add_2qb_gate method to take pulse_type as input, to add all parameters of the respective type
* add pulse_params to pulse types: this static method returns a dictionary of parameters of the pulse type with the respective initial values

MichKe95 committed:
*** Prepare branch for merge with Master ***

MichKe95 committed:
** Prepare for Pull Request **

* Clean up code
* Improve add_2qb_gate() by pulling the pulse function finding routine out of the for loop
* Include short documentation for device in the header

MichKe95 committed:
** Prepare for Pull Request **

* change Halfway to FLIP gate

MichKe95 committed:
** Prepare for Pull Request **

* change name of decay_freq_seq to T1_freq_sweep_seq to be coherent with other routines in code

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -266,10 +272,9 @@ def Ramsey_time_with_flux_seq(qb_name, hard_sweep_dict, operation_dict,
     return seq, np.arange(seq.n_acq_elements())
 
 
-def chevron_seqs(qbc_name, qbt_name, qbr_name, hard_sweep_dict, soft_sweep_dict,
```
qbr_name is needed inside the function

MichKe95 replied:
Agree!

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -434,7 +649,7 @@ def cz_bleed_through_phase_seq(phases, qb_name, CZ_pulse_name, CZ_separation,
     CZ_pulse1 = deepcopy(operation_dict[CZ_pulse_name])
     CZ_pulse_len = CZ_pulse1['pulse_length']
-    drag_pulse_len = deepcopy(X90_1['sigma']*X90_1['nr_sigma'])
+    drag_pulse_len = deepcopy(X90_1['sigma']* 90_1['nr_sigma'])
```
"90_1" looks wrong

MichKe95 replied:
Agree!

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -360,17 +375,83 @@ def fluxpulse_scope_sequence(
     flux_pulse['ref_pulse'] = 'FPS_Pi'
     flux_pulse['ref_point'] = 'middle'
-    flux_pulse['pulse_delay'] = -flux_pulse.get('buffer_length_start', 0)
+    flux_pulse_delays = -delays - flux_pulse.get('buffer_length_start', 0)
```
If the aim is to get rid of the buffer, why not just set the buffer length to zero in the deepcopied pulse?

MichKe95 replied:
This was the case before (in the master branch) with the line 
`flux_pulse['pulse_delay'] = -flux_pulse.get('buffer_length_start', 0)`

This line, however, has no effect, as the pulse lengths are overwritten by
`swept_pulses = sweep_pulse_params(pulses,
                                      {'FPS_Flux.pulse_delay': -delays})`

One therefore has to first define the new delays `flux_pulse_delays` (reduced by the buffer start length) and then use them to sweep the pulse parameters with 
`swept_pulses = sweep_pulse_params(
        pulses, {'FPS_Flux.pulse_delay': flux_pulse_delays})`

chellings replied:
I meant:
flux_pulse['buffer_length_start'] = 0

antsr replied:
If you set the buffer to zero, then the gaussian filtered rising edge will be cut off.

MichKe95 replied:
Yes, I think the way it is solved here is ok in my opinion.

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -548,8 +765,13 @@ def cphase_seqs(qbc_name, qbt_name, hard_sweep_dict, soft_sweep_dict,
-    delay = max_flux_length + flux_pulse.get('buffer_length_start', 0) + \
-        flux_pulse.get('buffer_length_end', 0)
+    nr_flux_buffer = 4 if flux_pulse['pulse_type']=='BufferedNZHalfwayPulse' \
+        else 2 #for NZ pulse we have four buffers
```
add comment to make clear that this only applies to FLIP gates

MichKe95 replied:
Done!

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -548,8 +765,13 @@ def cphase_seqs(qbc_name, qbt_name, hard_sweep_dict, soft_sweep_dict,
+        flux_pulse.get('buffer_length_start', 0) +
+        flux_pulse.get('buffer_length_end', 0) +
+        nr_flux_buffer*flux_pulse.get('flux_buffer_length', 0) +
+        nr_flux_buffer*flux_pulse.get('flux_buffer_length2', 0))
```
- add comment to make clear that this only applies to FLIP gate.
- the parameter names "flux_buffer_length" and "flux_buffer_length2" are not really self-explanatory

MichKe95 replied:
Done!

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -559,9 +781,19 @@ def cphase_seqs(qbc_name, qbt_name, hard_sweep_dict, soft_sweep_dict,
     ro_pulses = generate_mux_ro_pulse_list([qbc_name, qbt_name],
                                             operation_dict)
 
+    ro_pulses[0]['ref_pulse'] = final_rotations[longer_pulse]['name']
```
would be cleaner to use multiref feature (as soon as @antsr has reviewed that other pull request ;-))

chellings replied:
This works after merging the latest master:

```
@@ -786,22 +786,16 @@ def cphase_seqs(qbc_name, qbt_name, hard_sweep_dict, soft_sweep_dict,
     #     print(f'delay adjusted to {delay*1e9:.2f} ns '
     #           f'to fulfill commensurability conditions with 16/2.4e9.')
     final_rotations[0]['ref_pulse'] = 'cphase_init_pi_qbc'
     final_rotations[0]['pulse_delay'] = delay
 
-    # make sure RO happens after last ge
-    ge_pulse_len = np.array(
-        [final_rotations[i]['sigma']*final_rotations[i]['nr_sigma']
-         for i in range(2)])
-
-    longer_pulse = ge_pulse_len.argmax()
-
-
     ro_pulses = generate_mux_ro_pulse_list([qbc_name, qbt_name],
                                             operation_dict)
 
-    ro_pulses[0]['ref_pulse'] = final_rotations[longer_pulse]['name']
+    # make sure RO happens after last ge
+    ro_pulses[0]['ref_pulse'] = [p['name'] for p in final_rotations]
 
     hsl = len(list(hard_sweep_dict.values())[0]['values'])
     params = {'cphase_init_pi_qbc.amplitude': np.concatenate(
         [initial_rotations[0]['amplitude']*np.ones(hsl//2), np.zeros(hsl//2)]),
               'cphase_final_pi_qbc.amplitude': np.concatenate(

```

MichKe95 replied:
Is now implemented as suggested!

chellings commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -299,19 +306,27 @@ def chevron_seqs(qbc_name, qbt_name, qbr_name, hard_sweep_dict, soft_sweep_dict,
+            (max_flux_length + flux_pulse.get('buffer_length_start', 0) + \
+            flux_pulse.get('buffer_length_end', 0) + \
+            2*flux_pulse.get('flux_buffer_length', 0) + \
+            2*flux_pulse.get('flux_buffer_length2', 0))
```
- add comment that this applies to FLIP only
- is the factor "2*" always correct (NZ vs unipolar?)

MichKe95 replied:
Indeed, this has to be identical with what is done for the cphase sequence! Is now fixed!

chellings commented on pycqed/measurement/waveform_control/pulse_library.py:
```diff
@@ -209,7 +209,7 @@ def chan_wf(self, chan, tvals):
         wave *= np.cos(
             2 * np.pi * (self.frequency * t_rel + self.phase / 360.))
-        return wave
+        return wave - wave[0]
```
I understand that the waveform should start with 0, but is it correct to shift the whole waveform? Shouldn't it rather just set wave[0] to 0?

MichKe95 replied:
For me both seem like legitimate solutions. 

I have to admit, however, that I did not see problematic behavior if the waveforms are non-zero in the beginning. I am not sure about the status, whether we still believe that non-zero waveforms might lead to malfunctioning of the 5014. May @antsr knows more here.

If this is the case, then we should fix all waveforms such that they do start with zero.

antsr replied:
It is an issue with the 5014, so waveforms need to start with 0. This should not have to be done on a pulse level though, but at the element level in the segment class

MichKe95 replied:
Then I will revert this change!

And is it already done like that in the segment class? Otherwise we should create an issue for that.

chellings commented:
Thank you for preparing this pull request, which is another important step towards S17. In addition to some specific comments I left, I would have two suggestions:
- (if not done yet) test on a real setup whether everything works well after the recent merge of the master and after the cleanup
- test on a real setup whether all modified measurements (see fluxing_sequences) still work as intended for conventional (non-FLIP) gates (maybe this has been tested on the branch that @jo3141 is using?)

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -313,7 +313,7 @@ def process_data(self):
                         'sweep_points'], qbn)} for qbn in self.qb_names}
             self.proc_data_dict['sweep_points_dict'] = sweep_points_w_calpts
-        except TypeError as e:
+        except (TypeError, AttributeError) as e:
```
in which case do you get attribute errors?

MichKe95 replied:
I have to admit, that I unfortunately do not remember. We might revert this change.

MichKe95 replied:
Change reverted.

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -329,8 +329,25 @@ def process_data(self):
         self.data_to_fit = self.get_param_value('data_to_fit')
-        if self.cal_states_rotations is not None:
+        if self.cal_states_rotations is not None \
+            and not self.metadata.get('global_PCA', False):
```
usually for analysis related parameters  we use get_param_value(), which is a wrapper that checks both in the metadata and the options_dict, which is quite handy to modify parameters of the analysis

MichKe95 replied:
Nice! I implemented it!

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -476,6 +493,17 @@ def rotate_data(meas_results_per_qb, channel_map,
                         cal_zero_points=cal_zero_points,
                         cal_one_points=cal_one_points)
+
+                # if no calibration points are used rotate data manually
+                if len(cal_states_dict[qb_name]) == 0:
+                    mean = np.mean(
+                        rotated_data_dict[qb_name][data_to_fit[qb_name]])
+                    middle = (np.max(
+                        rotated_data_dict[qb_name][data_to_fit[qb_name]]) +
+                              np.min(rotated_data_dict[qb_name][
+                                         data_to_fit[qb_name]])) / 2
+                    rotated_data_dict[qb_name][data_to_fit[qb_name]]*=np.sign(
+                                                                middle-mean)
```
pretty sure we had code to do that already somewhere, no? @stephlazar maybe knows best

MichKe95 replied:
I did not find it in the jungle of analysis version 2. ;-) 

@stephlazar might want to correct me.

nathlacroix replied:
Soon analysis v3 will fix everything 🚀

nathlacroix commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,991 @@
+
+class Device(Instrument):
+    def __init__(self, name, qubits, connections, **kw):
+        super().__init__(name, **kw)
```
Would be important to have a docstring here. What are connections? How are they represented?

MichKe95 replied:
Done!

nathlacroix commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,991 @@
+        qbc_name = qbc.name
+        qbt_name = qbt.name
+
+        if [qbc_name, qbt_name] not in self.connectivity_graph() and [qbt_name,
+                                                                      qbc_name] not in self.connectivity_graph():
+            raise ValueError('Qubits are not connected!')
```
I would suggest a wrapper to the device class to do this, it will be useful also at some other places and we might be able to build upon it when using swaps.:
``` 
def check_connection(self, qubit_a, qubit_b, connectivity_graph=None, raise_exception=True):

    if connectivity_graph is None:
        connectivity_graph = self.connectivity_graph()
    # convert qubit object to name if necessary
    if not isintance(qubit_a, str):
        qubit_a = qubit_a.name
    if not isinstance(qubit_b, str):
        qubit_b = qubit_b.name
    if [qubit_a, qubit_b] not in self.connectivity_graph() and 
        [qubit_b, qubit_a] not in self.connectivity_graph():
        if raise_exception:
            raise ValueError(f'Qubits {[qubit_a, qubit_b]}  are not connected!')
        else:
            log.warning('Qubits are not connected!')
            # TODO: implement what we want in case of swap (e.g. determine shortest path of swaps)
            
```

nathlacroix replied:
just misses a doc string ;)

MichKe95 replied:
Nice! Now implemented in the code!

nathlacroix commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,991 @@
+        if analyze_RB:
+            pla.PipelineDataAnalysis()
+
+    def measure_chevron(self, qbc, qbt, hard_sweep_params, soft_sweep_params,
```
Merging all these functions in the master will require extra care, because some of those have been modified in the master branch which you branched off a long time ago, but they won't cause any merge conflict since you are writing them in a new file. To avoid introducing again all the bugs that have been fixed, the last version of these functions currently in the master (valid also for measure cphase, etc.) should be taken, which is currently not the case.

chellings replied:
Thank you, Nathan, for pointing that out. It raises another question: there is a branch enh/cphase_dynphase with many further improvements to cphase and dynphase measurements. If Michael has to copy new versions of the functions anyways, would it make sense to first process a pull request from the other branch to have the most recent versions in the master, and then copy (and adapt) the most recent versions afterwards? Otherwise we do the same work twice.

And why are the functions not removed from mqm if they are now here? Is that intended for backward compatibility? If yes, the functions in mqm should be replaced by wrappers that create a device object and call the new functions. Otherwise, if we do not need this backward compatibility, the functions in mqm should be removed.

MichKe95 replied:
You mean specifically for the methods in the multi_qubit_module? It should not require too much work to compare them with the ones in the master branch.

MichKe95 replied:
Yes, I will then wait with adapting the routines until the other PR is finished. Who is responsible for that one?

Concerning the mqm routines. I think we should delete them, as we should refrain from calling the method outside our device instance, which contains all device parameters.

MichKe95 replied:
As discussed with @chellings I updated the methods in the device class to use the current version in the multi qubit module of the master branch. 

I also deleted the version in the multiqubit module and have errors raised when the method is being called.

nathlacroix commented on pycqed/measurement/waveform_control/pulse.py:
```diff
@@ -131,3 +131,10 @@ def chan_wf(self, channel, tvals):
+        """
+        Returns a dictionary of pulse parameters and initial values.
+        """
+        raise NotADirectoryError('pulse_params() not implemented for your pulse')
```
why use this exception? I guess NotImplementedError was meant here.

MichKe95 replied:
Oops, yes auto-complete set me a trap ;-)

nathlacroix reviewed (changes requested):
Nice work: I think this device class will be important for future development; thanks for kickstarting it, and I would be really happy to see it soon integrated into the master. I think there remain a few things to change and also some points to keep in mind when merging into the master, as this is a quite large pull request. See dedicated comments.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r427149857

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r427156462

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r427164792.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r427165497.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r427170165.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r429177654.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r429190281.

MichKe95 committed:
Fix bug in get_pulse_par.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r429195123.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r429180687.

antsr commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -329,8 +329,25 @@ def process_data(self):
             self.cal_states_analysis()
+        elif self.get_param_value('global_PCA', False):
+            self.cal_states_dict_for_rotation = {qb: [] for qb in self.qb_names}
+            if self.get_param_value('TwoD', default_value=False):
```
why pass default here as keyword argument and as positional argument before. Inconsistent style might create confusion

MichKe95 replied:
Agree! Now pass in default value as kwarg , as is the case for the other instances where we call this method.

antsr commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,1032 @@
+                           parameter_class=ManualParameter)
+
+        self.add_parameter('operations',
+                           docstring='a list of all operations available on the qubit',
```
The docstring seems a bit imprecise. Its operations on the device, right?

I think that the difference between `device.operations()` and `device.get_operation_dict()` should be more explicitly explained in the docstrings and the difference should also ideally be reflected in the function names.

MichKe95 replied:
Yes, the doctoring was adapted to be more descriptive. 

Concerning the naming: I adapted the conventions for the QudevTransmon to the device object for ease of use and I would stick to that.

antsr commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,1032 @@
+                           docstring="Stores the connections between the qubits "
+                                     "in form of a list of lists [qbi, qbj]",
+                           parameter_class=ManualParameter,
+                           initial_value=connectivity_graph
```
why pass in the connections as pairs of qubit objects, if you are only going to store them as pairs of qubit names. I think it would make sense to omit the init argument or to make it the same format as this parameter.

Also the docstring should indicate that it stores qubit names, not instances.

MichKe95 replied:
Agree. Now, allow for both qubit names or object for both the qubits, as well as the connectivity_graph input.

nathlacroix replied:
I think introducing this connectivity graph is really a good feature :) Though I've more often seen directed acyclic graphs stored in python as vertices as keys, and neighbors a list of neighbors as values,
see e.g. https://www.python.org/doc/essays/graphs/. For 3 qubits in a chain it would be {"qb1": ["qb2"], "qb2": ["qb1", "qb3"], "qb3": ["qb2"]}
Using that representation, we cal also easily find paths through non-neighboring qubits (see attached link), and I think it could become useful when doing swaps based on connectivity.

antsr commented on pycqed/measurement/waveform_control/pulse.py:
```diff
@@ -131,3 +131,10 @@ def chan_wf(self, channel, tvals):
                                   .format(str(type(self))[1:-1]))
+
+    @staticmethod
+    def pulse_params():
```
I think this should be a classmethod instead. (`@classmethod` with first argument `cls`, similar to `self`)

It would be nice to add this for other existing pulses aswell, such that they can also be used with the device.

MichKe95 replied:
Done!

antsr reviewed (changes requested):
This pull request is quite large, and hard to very thoroughly review, but I guess we should still merge it after the open comments are addressed, to be able to move on. The concepts can then be incrementally improved as they are used.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62/files/549288298de9057a902c1fcebbe027c62c8c8455#r430299049.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r430646088.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r430590496.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r430600971

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r430604921

Now, allow for both qubit names or object for both the qubits, as well as the connectivity_graph input.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r430649409

MichKe95 committed:
Compare measure_chevron with version in multi qubit module of master branch.

MichKe95 committed:
Compare measure_cphase with version in multi qubit module of master branch.

MichKe95 committed:
Compare measure_dynamic_phase with version in multi qubit module of master branch.

MichKe95 committed:
Compare measure_two_qubit_randomized_benchmarking with version in multi qubit module of master branch.

MichKe95 committed:
Compare measure_tomography with version in multi qubit module of master branch.

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r429177068.

MichKe95 committed:
Raise error when trying to using mqm methods that have been implemented for the device object.

MichKe95 commented:
I am now done implementing your comments! Thanks a lot for your reviewing.

If @antsr or @nathlacroix want to review the changes, you can go ahead and do so. Otherwise I might go ahead and merge the pull request!

chellings commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,1012 @@
+                                            qbc, qbt, 'basis_rotation')()
+
+        qbc_name = qbc.name
+        qbt_name = qbt.name
```
delete these two unused variables

MichKe95 replied:
Addressed in https://github.com/QudevETH/PycQED_py3/pull/62/commits/90cdd02cd058de89758a207786f37740edd93bbf

chellings commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,1012 @@
+            qbt (QuDev_transmon): target qubit / non-fluxed qubit
+        '''
+        qbc_name = qbc.name
+        qbt_name = qbt.name
```
delete these two unused variables

MichKe95 replied:
Addressed in https://github.com/QudevETH/PycQED_py3/pull/62/commits/90cdd02cd058de89758a207786f37740edd93bbf

chellings commented on pycqed/instrument_drivers/meta_instrument/device.py:
```diff
@@ -0,0 +1,1012 @@
+                fsqs.dynamic_phase_seq(
+                    qb_name=qb.name, hard_sweep_dict=hard_sweep_params,
+                    operation_dict=self.get_operation_dict(),
+                    cz_pulse_name=cz_pulse_name + f' {qbc.name} {qbt.name}',
```
Does this mean that the cz_pulse_name argument of the function is not the actual pulse name, but only the prefix? This seems to be different from the convention used in measure_cphase. Should be unified.

MichKe95 replied:
@chellings this was resolved by you, right?

chellings replied:
I have finally decided to ignore it for now because I found out that this inconsistency exists also in other modules. We might at some point clean up the naming scheme throughout all modules. To distinguish between the prefix and the full name, we could either use cz_pulse_prefix and cz_pulse_name or, e.g., cz_pulse_name and cz_operation (or cz_op_code).

MichKe95 committed:
* Fix bug in measure_cphase where name of cz_pulse was not correctly passed to the sequence cphase_seqs.
* device.qubits is now parameter containing the qubit objects.
* add device.qubit_names, a list of qubit names .
* all algorithms now accept both QudevTransmon objects as well as qubit names.
* Remove unused qbc_name and qbt_name in measure_dynamic_phases() and measure_cphase().

nathlacroix reviewed (commented):
Sounds good for me. I just have a minor comment on the connectivity graph, though not critical at this stage.

MichKe95 committed:
* Move device algorithms into multi_qubit_module.py.
* Create wrapper functions for frequently used device algorithms.

MichKe95 committed:
* Fix bug in check_connection to use connectivity_graph input.

chellings committed:
measure_cphase: changed order of parameters to be compatible with master branch, minor fix in docstring

MichKe95 committed:
Hot fix for get_pulse_par to now correctly return the parameter object!

chellings committed:
mqm.measure_cphase bugfix: wrong use of cz_pulse_name

antsr commented on pycqed/measurement/waveform_control/pulse_library.py:
```diff
@@ -63,6 +63,27 @@ def __init__(self, name, element_name, I_channel, Q_channel,
         self.alpha = alpha
         self.phi_skew = phi_skew
 
+    @classmethod
```
I don't like very much the necessity to add all these paraneters and defautls here, as it duplicates the information in the init function.

You can retrieve the arguments and their default values from the definition of the init function via:
```
import inspect
inspect.signature(cls.__init__)
```

This would allow defining this function only once in the base class

antsr replied:
Example:

```
class Foo:
    def __init__(self, a, *args, b=1, **kw):
        pass
    
import inspect
sig = inspect.signature(Foo.__init__)

print(sig.parameters.keys())

par = sig.parameters['a']
print(par.kind, par.default)

par = sig.parameters['b']
print(par.kind, par.default)

par = sig.parameters['args']
print(par.kind, par.default)

par = sig.parameters['kw']
print(par.kind, par.default)
```
Output:
> odict_keys(['self', 'a', 'args', 'b', 'kw'])
> POSITIONAL_OR_KEYWORD <class 'inspect._empty'>
> KEYWORD_ONLY 1
> VAR_POSITIONAL <class 'inspect._empty'>
> VAR_KEYWORD <class 'inspect._empty'>

MichKe95 replied:
Addressed by https://github.com/QudevETH/PycQED_py3/pull/62/commits/5094f90da89c37a43e150e21f4bff14c02e32497

chellings committed:
Merge branch 'qudev_master' into devel/halfway_gate

chellings committed:
device.get_operation_dict: do not add two copies of two-qb operations, but instead add two reference to the same operation

chellings committed:
Device.get_operation_dict: store op_code in operations
(just like QuDev_transmon.get_operation_dict does)

MichKe95 committed:
Clean up Pulse Library

* remove BufferedCZPulseEffectiveTime pulse
* implement loading of default parameters from pulse_params in Pulse init method. This avoids duplication of code
* have the basis_rotation parameter set device method, as this parameter is not an attribute of the pulse object

MichKe95 committed:
* if basis_rotation part of pulse attributes, do not overwrite default value set by pulse_params

MichKe95 committed:
* move flipping sign of data into correct module

MichKe95 committed:
Hot fix: instantiate channels as [] not as None to avoid error in init script

MichKe95 committed:
Merge branch 'qudev_master' into devel/halfway_gate

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py

MichKe95 committed:
* Fix analysis after merge

MichKe95 committed:
* Fix analysis after merge

nathlacroix reviewed (approved):
I looked at changes related to tda and analysis toolbox. Looks good to me, thanks for adapting to the new master.
You could maybe just add a docstring in a_tools for your keyword but you already commented the code, so that's probably ok for now. I would merge as soon as possible :)
Thanks for this great pull request! Hope most things will work as intended :)

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -2017,45 +2027,64 @@ def calibrate_n_qubits(qubits, f_LO, sweep_points_dict, sweep_params=None,
+
+    if qbr is None:
+        qbr = qbt
+    elif qbr != qbc and qbr != qbt:
```
```
    elif isinstance(qbr, str):
        qbr = dev.get_qb(qbr)
    if qbr != qbc and qbr != qbt:
```

MichKe95 replied:
Done!

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -2017,45 +2027,64 @@ def calibrate_n_qubits(qubits, f_LO, sweep_points_dict, sweep_params=None,
 
-    channels_to_upload = [qbc.get_operation_dict()[cz_pulse_name]['channel']]
+    # sweep over flux pulse amplitude of qbc
+    channels_to_upload = [qbc.flux_pulse_channel()]
```
Always correct to upload only one channel? What if cz_pulse_name is set to a FLIP gate? Couldn't the user then decide to also put parameters of the other fluxed qubits into the sweep_points?

MichKe95 replied:
Agree! Add new channels_to_upload parameter, that contains the names of the channels that should be reuploaded during the soft sweep and which is set to the flux pulse channels of qbt and qbc by default.

chellings reviewed (approved):
Thank you for all your effort. I am happy that we can now merge this. Ants told me that we do not need to wait for a final approval from his side. Please have a short look at my two final comments before you merge. Thank you!

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r436256481

MichKe95 committed:
Address https://github.com/QudevETH/PycQED_py3/pull/62#discussion_r436256834

MichKe95 commented:
Thanks for all your inputs, I will proceed with the merge!
